nni/contrib/compression/base/config.py

```diff
@@ -0,0 +1,157 @@
+        if 'exclude' in config:
+            config.pop('exclude')
+            ex_op_types.update(config.get('op_types'))
+            ex_op_names.update(config.get('op_names'))
+            ex_op_names_re.update(config.get('op_names_re'))
+    if ex_op_types or ex_op_names or ex_op_names_re:
+        for config in config_list:
+            config['exclude_op_types'] = list(ex_op_types)
+            config['exclude_op_names'] = list(ex_op_names)
+            config['exclude_op_names_re'] = list(ex_op_names_re)
+
+    # trans pruning config
+    group_id_candidate = 0
+    for config in config_list:
+        sparse_ratio = None
+        group_id = None
+        max_sparse_ratio = config.pop('max_sparsity_per_layer', None)
+        if 'sparsity_per_layer' in config or 'sparsity' in config:
+            sparse_ratio = config.pop('sparsity_per_layer', config.pop('sparsity'))
+        if 'total_sparsity' in config:
+            sparse_ratio = config.pop('total_sparsity')
+            group_id = group_id_candidate
+            group_id_candidate += 1
+        if sparse_ratio is not None:
+            config['target_names'] = ['weight', 'bias']
+            config['target_settings'] = {
+                'weight': {
+                    'sparse_ratio': sparse_ratio,
+                    'max_sparse_ratio': max_sparse_ratio,
+                    'global_group_id': group_id,
+                    'sparse_granularity': 'default',
+                }
+            }
+
+    # trans quantization part
+    for config in config_list:
+        target_names = []
+        target_settings = {}
+        if 'quant_types' in config:
+            quant_types, quant_bits = config.pop('quant_types'), config.pop('quant_bits')
+            if 'input' in quant_types:
+                target_names.append(INPUT_PREFIX)
+                quant_bit = quant_bits if isinstance(quant_bits, int) else quant_bits['input']
+                target_settings[INPUT_PREFIX] = {'quant_dtype': f'int{quant_bit}'}
+            if 'weight' in quant_types:
+                target_names.extend(['weight', 'bias'])
+                quant_bit = quant_bits if isinstance(quant_bits, int) else quant_bits['weight']
+                target_settings['weight'] = {'quant_dtype': f'int{quant_bit}'}
+                target_settings['bias'] = {'quant_dtype': f'int{quant_bit}'}
+            if 'output' in quant_types:
+                target_names.append(OUTPUT_PREFIX)
+                quant_bit = quant_bits if isinstance(quant_bits, int) else quant_bits['output']
+                target_settings[OUTPUT_PREFIX] = {'quant_dtype': f'int{quant_bit}'}
+            config['target_names'] = target_names
+            config['target_settings'] = target_settings
+    return config_list
+
+
+def select_modules_by_config(model: torch.nn.Module, config: Dict[str, Any]) -> Tuple[Dict[str, torch.nn.Module], Dict[str, Any]]:
+    """
```

Comment:
Better to add more explanation on how this "selection" works.

Reply:
add comments

```
    This is a helper function for selecting the modules in model specified in config.

    There are six optional keys in config to specify the module,
    note that the module name should be the same as the name obtained from ``model.named_modules()``,
    module type is the ``__name__`` of the module class:
        - ``op_names``: a module name list, the modules with these names will be selected.
        - ``op_types``: a module type name list, the modules satisfied these types will be selected.
        - ``op_names_re``: a regular expression list, the modules satisfied the regular expressions will be selected.
        - ``exclude_op_names``: a module name list, the modules with these names will be excluded.
        - ``exclude_op_types``: a module type name list, the modules satisfied these types will be excluded.
        - ``exclude_op_names_re``: a regular expression list, the modules satisfied the regular expressions will be excluded.

    A module is selected if it satisfies all the following conditions:
        1. If ``op_names`` or ``op_names_re`` is not empty, the module name should in ``op_names``
           or satisfied one of regular expressions in ``op_names_re``.
        2. If ``op_types`` is not empty, the module type name should in ``op_types``.
        3. If ``exclude_op_names`` or ``exclude_op_names_re`` is not empty, the module name should not in ``exclude_op_names``
           and should not satisfied each of regular expressions in ``exclude_op_names_re``.
        4. If ``exclude_op_types`` is not empty, the module type name should not in ``exclude_op_types``.

```